crud/common/sharding_key.lua

```diff
@@ -0,0 +1,122 @@
+local fiber = require('fiber')
+local errors = require('errors')
+local vshard = require('vshard')
+
+local const = require('crud.common.const')
+local dev_checks = require('crud.common.dev_checks')
+local cache = require('crud.common.sharding_key_cache')
+local utils = require('crud.common.utils')
+
+local FetchShardingKeyError = errors.new_class('FetchShardingKeyError', {capture_stack = false})
+local GetShardingKeyFieldnosError = errors.new_class('GetShardingKeyFieldnosError',  {capture_stack = false})
+
+local FETCH_FUNC_NAME = '_crud.fetch_on_storage'
+
+local sharding_key_module = {}
+
+-- Build a structure similar to index, but it is not a real index object,
+-- it contains only parts key with fieldno's.
+function sharding_key_module.as_index_object(space_format, space_name, sharding_key_def)
+    dev_checks('table', 'string', 'table')
+
+    local fieldnos = {}
+    local fieldno_map = utils.get_format_fieldno_map(space_format)
+    for _, field_name in ipairs(sharding_key_def) do
+        local fieldno = fieldno_map[field_name]
+        if fieldno == nil then
+            return nil, GetShardingKeyFieldnosError:new(
+                "No such field (%s) in a space format (%s)", field_name, space_name)
+        end
+        table.insert(fieldnos, {fieldno = fieldno})
+    end
+
+    return {parts = fieldnos}
+end
+
+-- It's a public method to access it in tests.
+function sharding_key_module.fetch_on_storage()
+    local sharding_key_def = {}
+    local sharding_key_space = box.space._ddl_sharding_key
+    if sharding_key_space == nil then
+        return sharding_key_def
+    end
+
+    local SPACE_NAME_FIELDNO = 1
+    local SPACE_SHARDING_KEY_FIELDNO = 2
+    for _, tuple in ipairs(sharding_key_space:select()) do
+        local space_name = tuple[SPACE_NAME_FIELDNO]
+        local key = tuple[SPACE_SHARDING_KEY_FIELDNO]
+        sharding_key_def[space_name] = key
+    end
+
+    return sharding_key_def
+end
+
+local sharding_key_fetch_in_progress = false
+local sharding_key_fetch_cond = fiber.cond()
+
+-- Under high load we may get a case when many fibers will fetch sharding keys
+-- from storages. It is not good from performance point of view.
+-- concurrent_fetch() wraps a fetch_on_storage() to limit a number of fibers
+-- that fetches a sharding key by a single one, other fibers will wait for a
+-- FETCH_SHARDING_KEY_TIMEOUT and then make one more attempt to fetch sharding
+-- key if a cache is still empty.
+local function concurrent_fetch(replicaset)
+    dev_checks('table')
+
+    if sharding_key_fetch_in_progress == true and
+       not sharding_key_fetch_cond:wait(const.FETCH_SHARDING_KEY_TIMEOUT) then
+        return nil, FetchShardingKeyError:new(
+                        'Waiting for sharding key to be fetched is timed out')
+    end
+
+    local sharding_key_def = {}
+    if sharding_key_fetch_in_progress == false then
```

Comment:
If the variable is `true` (another fiber gets the lock), we'll just return `{}`.

Reply:
rewrote `reentrant_fetch()`